Bump d3-scale v1.0.7 to v2.2.2

watson opened the pull request:
### Note to reviewers

Besides updating the modules, I had to also update some of the code to force the type of a return value due to changes in the types:

In reality `xScale` can return either a number or `undefined`. I've not been able to figure out under which conditions it can return `undefined` but I assume it's if the input in invalid.

The `undefined` return value was added to `@types/d3-scale` for the v2.2 release in this PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/48095

Commit:
Bump d3-scale to v2.2.2

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/)
* Commit: ffccf7f691c646515655c37b31df3dca3614c216
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Build Kibana](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/execution/node/370/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @watson

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/","number":"137258","commit":"ffccf7f691c646515655c37b31df3dca3614c216"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/)
* Commit: ffccf7f691c646515655c37b31df3dca3614c216
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Build Kibana](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/execution/node/370/log)
* [API Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/execution/node/380/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @watson

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/","number":"137258","commit":"ffccf7f691c646515655c37b31df3dca3614c216"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/140963/)
* Commit: 1ad65e0ac7319232c2efd98e5cf9df5a66616fec
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [API Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/140963/execution/node/376/log)

### History
* :broken_heart: [Build #137258](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/) failed ffccf7f691c646515655c37b31df3dca3614c216

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @watson

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137258/","number":"137258","commit":"ffccf7f691c646515655c37b31df3dca3614c216"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/140963/","number":"140963","commit":"1ad65e0ac7319232c2efd98e5cf9df5a66616fec"}]}
PIPELINE-->

watson commented:
While we want to backport this to v6.8, I had to remove the label, as it would never work. This is because v6.8 still runs on Node.js 10 and ESM modules are not supported in that version of Node.js (d3-color v3 is pure ESM).

rashmivkulkarni commented:
do u mind adjusting the labels @watson , I see this is still a draft PR, but has 7.17.1 label which is going out today. Thanks.

Commit:
Expect xScale to return a number

In reality xScale can return either a number or `undefined`. I've not
been able to figure out under which conditions it can return `undefined`
but I assume it's if the input in invalid.

The `undefined` return value was added to `@types/d3-scale` for the v2.2
release in this PR:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/48095

watson commented:
@elastic/ml-ui @elastic/infra-monitoring-ui I could use both of your help with this PR:

After upgrading `@types/d3-scale` a bunch of TypeScript tests failed. I tried to deal with some of them belonging to @elastic/ml-ui (but not very well) and the ones belonging to @elastic/infra-monitoring-ui requires more knowledge of the `infra` plugin than I have.

So if you have time, I'd love your help ☺️

@elastic/ml-ui: You probably just need to verify that my last commit is correct.
@elastic/infra-monitoring-ui: I hope you can guide me through creating a new commit to fix the TS issues/push a commit yourself.

walterra reviewed (approved):
Thanks for the ping! I looked at the ML related changes, did a local test, also read through some related discussion on definitely typed where they consider this an edge case as well, so LGTM, thanks for the update! 👍

weltenwort commented:
I'll take a look at the `infra` type errors

weltenwort committed:
Add fallbacks for extra-domain values

weltenwort committed:
Merge branch 'main' into bump-d3-color

weltenwort committed:
Add more fallback values

weltenwort commented:
@watson I fixed the type errors in the `infra` plugin. The remaining errors seem to be unrelated and in `apm` and `security_solution`.

watson commented:
@elastic/apm-ui I forgot to ping you on this issue previously. I could use your help with this PR:

After upgrading `@types/d3-scale` a bunch of TypeScript tests failed. I tried to deal with some of them, but could use your help with the ones belonging to your team - either to guide me through creating a new commit to fix the TS issues or push a commit yourself ❤️ 

So if you have time, I'd love your help ☺️

Commit:
Merge branch 'main' into bump-d3-color

Commit:
fix remaining type errors

weltenwort reviewed (approved):
`infra` plugin changes LGTM

YulNaumenko reviewed (approved):
Security Solution changes LGTM

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|7.17|**Backport failed because of merge conflicts**|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 104888
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
<!--- Backport version: 7.4.0 -->
